packages/aws-cdk/lib/settings.ts

```diff
@@ -29,42 +35,63 @@ export class Configuration {
    * Load all config
    */
   public async load(): Promise<this> {
-    await this.userConfig.load(PER_USER_DEFAULTS);
-    await this.projectConfig.load(DEFAULTS);
-    return this;
-  }
+    const userConfig = await loadAndLog(USER_DEFAULTS);
+    const projectConfig = await loadAndLog(PROJECT_CONFIG);
+    const projectContext = await loadAndLog(PROJECT_CONTEXT);
+
+    // If the config currently has a 'context' setting, use that as context.
+    // Otherwise, load context from a separate file.
+    this.contextIsInLegacyFile = projectConfig.get([CONTEXT_KEY]) !== undefined;
+    if (this.contextIsInLegacyFile) {
+      if (!projectContext.empty) {
+        // Let's check context is not in two places, that's just a recipe for
+        // disaster while we're in the cutover period.
+        warning(`Found context in both ${PROJECT_CONFIG} and ${PROJECT_CONTEXT}; please remove one of them (${PROJECT_CONTEXT} will be ignored).`);
```

Comment:
I actually think that as long as the keys do not conflict we *should* support putting context in two places. I see `cdk.json` as user authored, and they can use it for application-specific context and `cdk.context.json` (maybe the name is not good) is the one auto-generated by the toolkit and includes a cache of the environmental context.

Reply:
I guess `cdk.context.json` is fine, because it interplays with the "cdk context" command.

Also, I would add a notation for comments in these JSON files so we can add some comment into the generated file with some hints on WTF is it and what users should do with it:

```json
{
  "$comment": "this is a generated file bla bla, dont edit by hand, use cdk.json for user-defined, commit it to store control, see <ref>",
  "context": {
    //
  }
}
```

Reply:
Having context in two places fucks with the "automatically detect where to save and clear values" algorithm though, especially if there's context in `cdk.json` which SHOULD have been in `cdk.context.json` because it's autogenerated.

Reply:
But that's just a migration/backwards-compatibility problem, correct? Can we always save environment context to `cdk.context.json` and remove it from `cdk.json` if found there?

Reply:
When you say "environment context" in this case you mean autogenerated context written by providers, right?

Reply:
yes